src/cryptonote_core/miner.cpp

```diff
@@ -411,5 +449,294 @@ namespace cryptonote
+    uint64_t total_sleep_time = 0;
+
+    uint64_t prev_total_time, current_total_time;
+    uint64_t prev_idle_time, current_idle_time;
+    uint64_t previous_process_time = 0, current_process_time = 0;
+
+    // get some starting data, then wait one windows
+    if(!get_system_times(prev_total_time, prev_idle_time))
+    {
+      LOG_ERROR(__func__ << " : get_system_times call failed, background mining will NOT work!");
+      return false;
+    }
+
+    while(!m_stop)
+    {
+      misc_utils::sleep_no_w(SLEEP_CYCLE); // zzZZzzz
+      if( !m_is_background_mining_enabled ) continue;
+      total_sleep_time += SLEEP_CYCLE; // at least ..
+
+      if( m_is_background_mining_started )
+      {
+        // figure out if we need to stop, and monitor mining usage
+
+        // previous is uninitialized
+        if(previous_process_time == 0)
+        {
+          // get some starting data, then wait one windows
+          if(!get_system_times(prev_total_time, prev_idle_time))
+            LOG_ERROR(__func__ << " : get_system_times call failed");
+
+          if(!get_process_time(previous_process_time))
+            LOG_ERROR(__func__ << " : get_process_time call failed!");
+
+          continue;
+        }
+
+        // has enough time passed for analysis?
+        if( total_sleep_time < BACKGROUND_MINING_MINER_MONITOR_INVERVAL_IN_SECONDS * 1000 )
+          continue;
+
+        // yay, if we're here then enough time has passed since our last reading!
+
+        if(!get_system_times(current_total_time, current_idle_time))
+        {
+          LOG_ERROR(__func__ << " : get_system_times call failed");
+          continue;
+        }
+
+        if(!get_process_time(current_process_time))
+        {
+          LOG_ERROR(__func__ << " : get_process_time call failed!");
+          continue;
+        }
 
+        uint64_t total_diff = (current_total_time - prev_total_time);
+        uint64_t idle_diff = (current_idle_time - prev_idle_time);
+        uint64_t process_diff = (current_process_time - previous_process_time);
+
+        // using ceil here is giving me a little edge, but just be a good network citizen and mine!
+        uint8_t idle_percentage = (uint8_t)( ceil( (idle_diff * 1.f / total_diff * 1.f) * 100) );
+        uint8_t process_percentage = (uint8_t)( ceil( (process_diff * 1.f / total_diff * 1.f) * 100) );
+        MGINFO(__func__ << " : idle percentage is " << unsigned(idle_percentage) << "\%, miner percentage is " << unsigned(process_percentage) << "\%");
+
+        if( idle_percentage + process_percentage < get_idle_threshold() )
+        {
+          MGINFO(__func__ << " : cpu is " << unsigned(idle_percentage) << "% idle, idle threshold is " << unsigned(get_idle_threshold()) << "\%, background mining stopping, thanks for your contribution!");
+          m_is_background_mining_started = false;
+
+          // reset process times
+          previous_process_time = 0;
```

Comment:
Hmm, I guess I want to start tracking from when mining starts. Say I don't reset it when mining pauses, a bunch of time passes, mining resumes, process usage will be lower than what's going on because I'm taking into account all of the down time when mining wasn't happening?